doc/man/man4/TTY.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/TTY.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH TTY 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Drivers and Modules"
.\"
.\"
.SH NAME
.B TTY
\- @PACKAGE_TITLE@ (TTY)
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/stropts.h>
.br
.B #include <sys/termios.h>
.\"
.\"
.SH DESCRIPTION
.PP
.B TTY
is provided by the
.BR "@PACKAGE_TITLE@ " ( @PACKAGE@ - @VERSION@ )
add-on package for
.BR "Linux Fast-STREAMS" .
The
.B @PACKAGE@
package contains header files and the following
.IR STREAMS (4)
modules and drivers:
.\"
.SS Drivers
.PP
.B @PACKAGE@
provides the following
.I STREAMS
pseudo-device drivers:
.TP
.BR ptm (4)
.I STREAMS
Pseudo-Terminal Master Driver.
.TP
.BR pts (4)
.I STREAMS
Pseudo-Terminal Slave Driver.
.TP
.BR pty (4)
.I STREAMS
Pseudo-Terminal Master and Slave Drivers.
.\"
.SS Modules
.PP
.PP
.B @PACKAGE@
provides the following
.I STREAMS
pushable modules:
.TP
.BR pckt (4)
.I STREAMS
Packet Mode Module.
.TP
.BR ptem (4)
.I STREAMS
Pseudo-Terminal Emulation Module.
.TP
.BR ttcompat (4)
.I STREAMS
TTY Compatibility Module.
.TP
.BR ldterm (4)
.I STREAMS
Line Discipline for Terminals.
.\"
.SS Devices
.PP
.B @PACKAGE@
provides the following
.I STREAMS
device drivers:
.\"
.\"
.\".SH NOTICES
.\".PP
.\"
.\"
.SH DEVICES
.PP
.TP
.I /dev/ptyXX
.TP
.I /dev/ptmx
.TP
.I /dev/ptm
.TP
.I /dev/ptsXX
.\"
.\"
.SH MODULES
.PP
.TP
.RI \(lq pckt \(rq
.TP
.RI \(lq ptem \(rq
.TP
.RI \(lq ttcomapt \(rq
.TP
.RI \(lq ldterm \(rq
.\"
.\"
.SH "SEE ALSO"
.PP
.BR grantpt (3tty),
.BR ptsname (3tty),
.BR unlockpt (3tty),
.BR ldterm (4),
.BR pckt (4),
.BR ptc (4),
.BR ptem (4),
.BR ptm (4),
.BR ptmx (4tty),
.BR pts (4tty),
.BR pty (4),
.BR ttcompat (4),
.BR streams_mkdev (8),
.BR test-tty (8).
.\"
.\"
.SH COMPATIBILITY
.PP
The
.B TTY
modules and drivers provided by the
.B @PACKAGE@
add-on package are compatible with
.I SVR 4.2 MP
and implementations based on
.I SVR 4.2 MP
including
.IR Solaris \(rg
and
.IR UnixWare \(rg,
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.IP \(em
.IP \(em
.IP \(em
.IP \(em
.\"
.\"
.\".SH CONFORMANCE
.\".PP
.\"
.\"
.SH HISTORY
.PP
A Terminal Subsystem based on
.I STREAMS
initially appeared in
.IR "SVR 4" .
.PP
.B @PACKAGE@
was started as a separate package from
.BR Linux Fast STREAMS
in August 2006.
.\"
.\"
.[
$LIST$
.]
.\"
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


isode/others/quipu/photo/tty.c
/*****************************************************************************

 @(#) $RCSfile$ $Name$($Revision$) $Date$

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2007  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date$ by $Author$

 -----------------------------------------------------------------------------

 $Log$
 *****************************************************************************/

#ident "@(#) $RCSfile$ $Name$($Revision$) $Date$"

static char const ident[] = "$RCSfile$ $Name$($Revision$) $Date$";

/* tty.c - display on any terminal */

#ifndef	lint
static char *rcsid =
    "Header: /xtel/isode/isode/others/quipu/photo/RCS/tty.c,v 9.0 1992/06/16 12:43:35 isode Rel";
#endif

/* 
 * Header: /xtel/isode/isode/others/quipu/photo/RCS/tty.c,v 9.0 1992/06/16 12:43:35 isode Rel
 *
 *
 * Log: tty.c,v
 * Revision 9.0  1992/06/16  12:43:35  isode
 * Release 8.0
 *
 */

/*
 *				  NOTICE
 *
 *    Acquisition, use, and distribution of this module and related
 *    materials are subject to the restrictions of a license agreement.
 *    Consult the Preface in the User's Manual for the full terms of
 *    this agreement.
 *
 */

#include "stdio.h"
#include "quipu/photo.h"

#define GREYSCALE "#OI&\\*o=_\"-;:,.  "

char mapping[16];
char *strcpy();

char buffer[500][500];
int lineno = 0, pos = 0;
extern int PIC_LINESIZE;
extern int NUMLINES;
int scale = 8;
int invert = 0;
int equal = 0;
int ln = 0;
int scalediv = 0;
int edges = 0;

char level[128];
char show[128];
char display[50][50];

/* ARGSUSED */
photo_start(name)
	char *name;
{
	char *ptr, *getenv();

	if ((ptr = getenv("photo_invert")) != (char *) NULL)
		if (strcmp(ptr, "true") == 0)
			invert = 1;

	if ((ptr = getenv("photo_equal")) != (char *) NULL) {
		if (strcmp(ptr, "true") == 0)
			equal = 1;
		else if (strcmp(ptr, "edge") == 0) {
			int i, j;

			for (i = 1; i < PIC_LINESIZE; i++)
				for (j = 1; j < NUMLINES; j++)
					buffer[i][j] = 0;
			edges = 1;
		}
	}
	if ((ptr = getenv("photo_mapping")) != (char *) NULL)
		(void) strcpy(mapping, ptr);
	else
		(void) strcpy(mapping, GREYSCALE);

	if ((ptr = getenv("photo_scale")) != (char *) NULL) {
		if (strcmp(ptr, "large") == 0)
			scale = 1;
		else if (strcmp(ptr, "small") == 0)
			scale = 8;
		else if (strcmp(ptr, "medium") == 0)
			scale = 4;
	}
	scalediv = scale * scale / 8;

	(void) printf("\n");
	return (0);
}

/* ARGSUSED */
photo_end(name)
	char *name;
{
	int i, j, k;
	int numlev;
	int total = 0;
	int totlevel = 0;
	int cnt;

	if (equal) {
		numlev = scale * scale * 2;
		for (i = 0; i < numlev; i++)
			level[i] = 0;

		for (i = 0; i < ln; i++)
			for (j = 0; j < PIC_LINESIZE / scale; j++)
				level[display[i][j]]++;

		for (i = 0; i < numlev; i++)
			totlevel += level[i];

		for (i = 0; i < numlev; i++) {
			total += level[i];
			show[i] = (total * 16) / totlevel;
			if (show[i] >= 16)
				show[i] = 15;
		}

		for (i = 0; i < ln; i++) {
			for (j = 0; j < PIC_LINESIZE / scale; j++) {
				if (invert)
					(void) putc(mapping[show[display[i][j]]], stdout);
				else
					(void) putc(mapping[15 - show[display[i][j]]], stdout);
			}
			(void) printf("\n");
		}
	} else if (edges) {
		/* edges by expansion */
		char ebuf[500][500];

		for (i = 1; i < PIC_LINESIZE; i++)
			for (j = 1; j < NUMLINES; j++)
				ebuf[i][j] = 0;

		for (i = 1; i < PIC_LINESIZE; i++)
			for (j = 1; j < NUMLINES; j++)
				if (buffer[i][j] == 1) {
					ebuf[i - 1][j - 1] = 1;
					ebuf[i - 1][j] = 1;
					ebuf[i - 1][j + 1] = 1;
					ebuf[i][j - 1] = 1;
					ebuf[i][j + 1] = 1;
					ebuf[i + 1][j - 1] = 1;
					ebuf[i + 1][j] = 1;
					ebuf[i + 1][j + 1] = 1;
				}
		for (i = 1; i < PIC_LINESIZE; i++)
			for (j = 1; j < NUMLINES; j++)
				if (buffer[i][j] == 1)
					ebuf[i][j] = 0;

		for (lineno = 0; lineno < NUMLINES; lineno += (2 * scale)) {
			for (k = 0; k < PIC_LINESIZE; k += scale) {
				cnt = 0;
				for (i = k; i < k + scale; i++)
					for (j = lineno; j < (2 * scale) + lineno; j++)
						cnt += ebuf[i][j];

/* Need to select a grey level on the strength of the edge
 *
 *				cnt *= 4;
 *				if (cnt == (scalediv * 16))
 *					cnt = (scalediv * 16) -1;
 *
 * Just set "strong" edge cells "on" for now ...
 */
				if (cnt > (scalediv * 2))
					cnt = (scalediv * 16) - 1;
				else
					cnt = 0;

				if (invert)
					(void) putc(mapping[(cnt / scalediv)], stdout);
				else
					(void) putc(mapping[15 - (cnt / scalediv)], stdout);
			}
			(void) putc('\n', stdout);
		}

	}
	return (0);

}

photo_black(length)
	int length;
{
	int i;

	if (scale == 1) {
		for (i = pos; i < length + pos; i++)
			if (invert)
				(void) putc(' ', stdout);
			else
				(void) putc('#', stdout);
		return;
	}
	for (i = pos; i < length + pos; i++)
		buffer[i][lineno] = 1;
	pos += length;
}

photo_white(length)
	int length;
{
	int i;

	if (scale == 1) {
		for (i = pos; i < length + pos; i++)
			if (invert)
				(void) putc('#', stdout);
			else
				(void) putc(' ', stdout);
		return;
	}
	for (i = pos; i < length + pos; i++)
		buffer[i][lineno] = 0;
	pos += length;
}

/* ARGSUSED */
photo_line_end(line)
	bit_string *line;
{
	int i, j, k, cnt;

	if (scale == 1) {
		(void) putc('\n', stdout);
		return;
	}
	lineno++;
	pos = 0;

	if (edges)
		return;

	if (lineno >= 2 * scale) {
		ln++;
		lineno = 0;
		for (k = 0; k < PIC_LINESIZE; k += scale) {
			cnt = 0;
			for (i = k; i < k + scale; i++)
				for (j = 0; j < 2 * scale; j++)
					cnt += buffer[i][j];

			if (equal) {
				display[ln][k / scale] = cnt;
				continue;
			}
			if (cnt == (scalediv * 16))
				cnt--;

			if (invert)
				(void) putc(mapping[cnt / scalediv], stdout);
			else
				(void) putc(mapping[15 - (cnt / scalediv)], stdout);
		}

		if (!equal)
			(void) putc('\n', stdout);
	}
}
